homeassistant/components/mqtt/mixins.py

```diff
@@ -863,12 +875,43 @@ async def async_removed_from_registry(self) -> None:
             self._cleanup_discovery_on_remove()
 
             # Clear the discovery topic so the entity is not rediscovered after a restart
-            await async_remove_discovery_payload(self.hass, self._discovery_data)
+            await async_remove_discovery_payload(
+                self.hass, cast(dict, self._discovery_data)
+            )
+
+    async def _async_clear_discovery_topic_if_entity_removed(
+        self,
+        hass: HomeAssistant,
+        discovery_data: dict[str, Any],
+        event: Event,
+    ) -> None:
+        """Clear the discovery topic if the entity is removed."""
+        if event.data["action"] == "remove":
+            # publish empty payload to config topic to avoid re-adding
+            discovery_hash: tuple[str, str] = discovery_data[ATTR_DISCOVERY_HASH]
+            hass.async_create_task(async_remove_discovery_payload(hass, discovery_data))
+            self._registry_hooks.pop(discovery_hash)()
 
     @callback
     def add_to_platform_abort(self) -> None:
         """Abort adding an entity to a platform."""
         if self._discovery_data:
+            discovery_hash: tuple = self._discovery_data[ATTR_DISCOVERY_HASH]
+            if (
+                self.registry_entry is not None
+                and discovery_hash not in self._registry_hooks
```

Comment:
If this can happen, we should handle it in `__init__` instead:
```py
if discovery_hash in self._registry_hooks:
    self._registry_hook.pop(discovery_hash)()
```

Reply:
Sharp